baseplate/core.py

```diff
@@ -123,6 +124,123 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+    the incoming requests.
+
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, secrets_store=None):
+        self.secrets = secrets_store
+
+    def make_context(self, token):
+        """:py:class:`AuthenticationContext` builder
+
+        :param token: JWT value originating from the Authentication service
+            either directly or from an upstream service
+        :rtype: :py:class:`AuthenticationContext`
+        """
+        return AuthenticationContext(token=token, secrets=self.secrets)
+
+
+class AuthenticationContext(object):
+    """Wrapper for the contextual authentication information
+
+    :param str token: the JWT value for the currently propagated authentication
+        context
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, token=None, secrets=None):
+        self.token = token
+        self.secrets = secrets
+        self.payload = {}
+        self._valid = None
+
+    def _secret(self):
+        if not self.secrets:
+            raise UndefinedSecretsException
+
+        return self.secrets.get_simple("jwt/authentication/secret")
+
+    @property
+    def valid(self):
+        """Validity of the current authentication context token
+
+        :type: bool
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and validated)
+
+        """
+        if self._valid is not None:
+            return self._valid
+        elif self.token is None:
+            return None
+
+        try:
+            self.payload = jwt.decode(self.token, self._secret(),
+                                      algorithms='RS256')
+            self._valid = True
+        except jwt.ExpiredSignatureError:
+            self._valid = False
+        except jwt.DecodeError:
+            self._valid = False
+
+        return self._valid
+
+    @property
+    def account_id(self):
+        """Authenticated account_id for the current authenticated context
+
+        :type: account_id string or None if context authentication is invalid
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`UndefinedAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        if self.token is None:
+            raise UndefinedAuthenticationError
+        if not self.valid:
+            return None
```

Comment:
Originally I was wrapping the payload access entirely to allow for the error-ing (and had the no token error raise in the `valid` method) but wanted to make the conditions for this error to be more granular.  Only the `sub` field of the payload is being tested here as opposed to other (future) fields that may come down the line which I don't know if they should throw errors until we define how they should be handled in a richer format (scopes and origin service have both been discussed) and I would probably think those would have different conditions for errors.  I am kind of torn with this being adaptable for future proofing versus this being strict to avoid contextual missteps in the future.